src/rgw/rgw_dedup_table.cc

```diff
@@ -0,0 +1,298 @@
+    occupied_count = 0;
+  }
+
+  //---------------------------------------------------------------------------
+  void dedup_table_t::remove_singletons_and_redistribute_keys()
+  {
+    for (uint32_t tab_idx = 0; tab_idx < entries_count; tab_idx++) {
+      if (!hash_tab[tab_idx].val.is_occupied()) {
+        continue;
+      }
+
+      if (hash_tab[tab_idx].val.is_singleton()) {
+        hash_tab[tab_idx].val.clear_flags();
+        redistributed_clear++;
+        continue;
+      }
+
+      const key_t &key = hash_tab[tab_idx].key;
+      uint32_t key_idx = key.hash() % entries_count;
+      if (key_idx != tab_idx) {
+        uint64_t count = 1;
+        redistributed_count++;
+        uint32_t idx = key_idx;
+        while (hash_tab[idx].val.is_occupied()   &&
+               !hash_tab[idx].val.is_singleton() &&
+               (hash_tab[idx].key != key)) {
+          count++;
+          idx = (idx + 1) % entries_count;
+        }
+
+        if (idx != tab_idx) {
+          if (hash_tab[idx].val.is_occupied() && hash_tab[idx].val.is_singleton() ) {
+            redistributed_clear++;
+          }
+          if (idx == key_idx) {
+            redistributed_perfect++;
+          }
+          hash_tab[idx] = hash_tab[tab_idx];
+          hash_tab[tab_idx].val.clear_flags();
+        }
+        else {
+          redistributed_loopback++;
+        }
+
+        redistributed_search_max = std::max(redistributed_search_max, count);
+        redistributed_search_total += count;
+      }
+      else {
+        redistributed_not_needed++;
+      }
+    }
+  }
+
+  //---------------------------------------------------------------------------
+  uint32_t dedup_table_t::find_entry(const key_t *p_key)
+  {
+    uint64_t count = 1;
+    uint32_t idx = p_key->hash() % entries_count;
+
+    // search until we either find the key, or find an empty slot.
+    while (hash_tab[idx].val.is_occupied() && (hash_tab[idx].key != *p_key)) {
+      count++;
+      idx = (idx + 1) % entries_count;
+    }
+    max = std::max(max, count);
+    total += count;
+    return idx;
+  }
+
+  //---------------------------------------------------------------------------
+  int dedup_table_t::add_entry(key_t *p_key, disk_block_id_t block_id, record_id_t rec_id,
+                               bool shared_manifest, bool valid_sha256)
+  {
+    value_t val(block_id, rec_id, shared_manifest, valid_sha256);
+    uint32_t idx = find_entry(p_key);
+    if (!hash_tab[idx].val.is_occupied()) {
+      if (occupied_count < entries_count) {
+        occupied_count++;
+      }
+      else {
+        return -EOVERFLOW;
+      }
+
+      hash_tab[idx].key = *p_key;
+      hash_tab[idx].val = val;
+      ldpp_dout(dpp, 20) << __func__ << "::add new entry" << dendl;
+      ceph_assert(hash_tab[idx].val.count == 1);
+    }
+    else {
+      ceph_assert(hash_tab[idx].key == *p_key);
+      hash_tab[idx].val.count ++;
+      if (!hash_tab[idx].val.has_shared_manifest() && shared_manifest) {
```

Comment:
what if either entries have `shared_manifest`?
can this happen?

Reply:
When an object is dedup both SRC and TGT are marked as having a shared-manifest and unless one of them is deleted then the next cycle going to load both with a shared-manifest flag